presto-main/src/main/java/io/prestosql/operator/DynamicFilterCollectOperator.java

```diff
@@ -0,0 +1,229 @@
+                    driverContext.addOperatorContext(operatorId, planNodeId, DynamicFilterCollectOperator.class.getSimpleName()),
+                    dynamicFilter,
+                    this);
+        }
+
+        @Override
+        public void noMoreOperators()
+        {
+        }
+
+        @Override
+        public OperatorFactory duplicate()
+        {
+            throw new UnsupportedOperationException("duplicate() is not supported for DynamicFilterCollectOperatorFactory");
+        }
+    }
+
+    private final OperatorContext context;
+    private boolean finished;
+    private Page current;
+    private final Consumer<Optional<TupleDomain<ColumnHandle>>> dynamicPredicateConsumer;
+    private final int maxRowCount;
+    private final long maxByteSize;
+
+    private final List<DynamicFilter.Descriptor> descriptors;
+    private ImmutableList.Builder<Page> pageBuilder;  // may be dropped if the predicate becomes too large.
+    private final PageChannelSelector selector;
+    private int positionsCount;
+    private long sizeInBytes;
+
+    public DynamicFilterCollectOperator(OperatorContext context, DynamicFilter dynamicFilter, Consumer<Optional<TupleDomain<ColumnHandle>>> dynamicPredicateConsumer)
+    {
+        this.context = requireNonNull(context, "context is null");
+        this.dynamicPredicateConsumer = dynamicPredicateConsumer;
+        this.maxRowCount = getDynamicFilteringMaxPerDriverRowCount(context.getSession());
+        this.maxByteSize = getDynamicFilteringMaxPerDriverByteSize(context.getSession()).toBytes();
+
+        this.descriptors = dynamicFilter.getDescriptors();
+        this.pageBuilder = new ImmutableList.Builder<>();
+        int[] buildChannels = descriptors.stream().mapToInt(d -> d.getBuildChannel()).toArray();
+        this.selector = new PageChannelSelector(buildChannels);
+        this.positionsCount = 0;
+        this.sizeInBytes = 0L;
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return context;
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return current == null && !finished;
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        verify(!finished, "DynamicFilterCollectOperator: addInput() may not be called after finish()");
+        current = page;
+        if (pageBuilder == null) {
+            return;  // the predicate became too large.
+        }
+
+        // Collect only the columns which are relevant for the JOIN.
+        Page projectedPage = selector.apply(page);
+        positionsCount += projectedPage.getPositionCount();
+        sizeInBytes += projectedPage.getSizeInBytes();
+        pageBuilder.add(projectedPage);
+
+        if (positionsCount > maxRowCount || sizeInBytes > maxByteSize) {
+            // The resulting predicate is too large, allow all probe-side values to be read.
+            dynamicPredicateConsumer.accept(Optional.empty());
```

Comment:
Why is this needed? I think we can simply skip calling `accept` with `Optional.empty()`

Reply:
Passing an `Optional.empty()` here helps the consumer to know that the collection failed early, so it would drop previously collected predicates:
https://github.com/prestosql/presto/blob/c343902c156401496bc89e6ca671a9747842594c/presto-main/src/main/java/io/prestosql/sql/planner/DynamicFilter.java#L66

Reply:
On the other hand, we can pass `TupleDomain.all()` (similar to how it's done by #1072).
Should be fixed at 1559eeb7b4.